src/Streamly/Internal/Data/Unfold/Enumeration.hs

```diff
@@ -0,0 +1,705 @@
+    step (x, stride, to) = return $
+        if stride == 0
+        then
+            if x <= to then Yield x $! (x, stride, to) else Stop
+
+        else
+            if stride > 0
+            then
+                if x <= to && x <= (maxBound :: a)
+                then Yield x $! (x + stride, stride, to)
+                else Stop
+            else
+                if x >= to && x >= (minBound :: a)
+                then Yield x $! (x + stride, stride, to)
+                else Stop
+
+-- Enumerate Unbounded Integrals ----------------------------------------------
+{-# INLINE enumerateFromIntegral #-}
+enumerateFromIntegral :: (Monad m, Integral a) => Unfold m a a
+enumerateFromIntegral = lmap (\a -> (a, a + 1)) enumerateFromStepIntegral
+
+{-# INLINE enumerateFromThenIntegral #-}
+enumerateFromThenIntegral :: (Monad m, Integral a ) => Unfold m (a, a) a
+enumerateFromThenIntegral = enumerateFromStepIntegral
+
+{-# INLINE enumerateFromToIntegral #-}
+enumerateFromToIntegral :: (Monad m, Integral a) => Unfold m (a, a) a
+enumerateFromToIntegral = lmap (\(a, c) -> (a, a + 1, c)) enumerateFromToStepIntegral
+
+{-# INLINE enumerateFromThenToIntegral #-}
+enumerateFromThenToIntegral :: (Monad m, Integral a) => Unfold m (a, a, a) a
+enumerateFromThenToIntegral = enumerateFromToStepIntegral
+
+-- Enumerate Bounded Integrals ------------------------------------------------
+{-# INLINE enumerateFromIntegralBounded #-}
+enumerateFromIntegralBounded :: (Monad m, Bounded a, Integral a) =>
+    Unfold m a a
+enumerateFromIntegralBounded =
+    lmap (\a -> (a, a + 1, maxBound)) enumerateFromToStepIntegralBounded
+
+{-# INLINE enumerateFromThenIntegralBounded #-}
+enumerateFromThenIntegralBounded :: (Monad m, Bounded a, Integral a ) =>
+    Unfold m (a, a) a
+enumerateFromThenIntegralBounded = enumerateFromStepIntegralBounded
+
+{-# INLINE enumerateFromToIntegralBounded #-}
+enumerateFromToIntegralBounded :: (Monad m, Integral a, Bounded a) =>
+    Unfold m (a, a) a
+enumerateFromToIntegralBounded =
+    lmap (\(a, c) -> (a, a + 1, c)) enumerateFromToStepIntegralBounded
+
+{-# INLINE enumerateFromThenToIntegralBounded #-}
+enumerateFromThenToIntegralBounded :: (Monad m, Integral a, Bounded a) =>
+    Unfold m (a, a, a) a
+enumerateFromThenToIntegralBounded = enumerateFromToStepIntegralBounded
+
+------------------------------------------------------------------------------
+-- Enumeration of Fractionals
+------------------------------------------------------------------------------
+
+{-# INLINE_NORMAL enumerateFromFractional #-}
+enumerateFromFractional :: (Monad m, Fractional a) => Unfold m a a
+enumerateFromFractional = enumerateFromNum
+
+{-# INLINE_NORMAL enumerateFromThenFractional #-}
+enumerateFromThenFractional :: (Monad m, Fractional a) => Unfold m (a, a) a
+enumerateFromThenFractional = enumerateFromThenNum
+
+{-# INLINE_NORMAL enumerateFromToFractional #-}
+enumerateFromToFractional :: (Monad m, Fractional a, Ord a) =>
```

Comment:
This and the next one are wrong as they won't round the upper limit to the nearest integer.

Reply:
Ah, ok the `enumerateFromToNum` is actually fractional and correctly implemented for fractionals. This is confusing because of incorrect naming.